### PR TITLE
[Spark] Fix time option evaluation

### PR DESCRIPTION
**Pull request by cloud-fan**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

We made a fix to evaluate current datetime functions for Delta options. However, the fix is not completed as it doesn't handle data types other than `TimestampType`. This PR fixes it.

## How was this patch tested?

A new test
## Does this PR introduce _any_ user-facing changes?

Yes, before the fix, Delta throws `[INTERNAL_ERROR] Cannot evaluate expression`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @scottsand-db

File: spark/src/main/scala-spark-master/shims/DeltaTableValueFunctionsShims.scala [hunk 1]
```diff
@@ -29,7 +29,7 @@ object DeltaTableValueFunctionsShims {
    * In Spark 4.0, SPARK-46331 [https://github.com/apache/spark/pull/44261] removed CodegenFallback
    * from a subset of DateTime expressions, making the `now()` expression unevaluable.
    */
-  def evaluateTimeOption(value: Expression): String = {
+  def evaluateOptionExpr(value: Expression): String = {
```
Comment (scottsand-db) [inline spark/src/main/scala-spark-master/shims/DeltaTableValueFunctionsShims.scala:32]: Hm. Is this not still time-related? There is `ComputeCurrentTime` below

Comment (cloud-fan) [reply spark/src/main/scala-spark-master/shims/DeltaTableValueFunctionsShims.scala:32]: you are right, we don't need to rename it.

File: spark/src/main/scala-spark-3.5/shims/DeltaTableValueFunctionsShims.scala [hunk 2]
```diff
@@ -26,7 +26,7 @@ object DeltaTableValueFunctionsShims {
    * In Spark 4.0, SPARK-46331 [https://github.com/apache/spark/pull/44261] removed CodegenFallback
    * from a subset of DateTime expressions, making the `now()` expression unevaluable.
    */
-  def evaluateTimeOption(value: Expression): String = {
+  def evaluateOptionExpr(value: Expression): String = {
```
Comment (cloud-fan) [inline spark/src/main/scala-spark-3.5/shims/DeltaTableValueFunctionsShims.scala:29]:
```suggestion
  def evaluateTimeOption(value: Expression): String = {
```

File: spark/src/main/scala-spark-master/shims/DeltaTableValueFunctionsShims.scala (hunk 1 above)
Comment (cloud-fan) [inline spark/src/main/scala-spark-master/shims/DeltaTableValueFunctionsShims.scala:32]:
```suggestion
  def evaluateTimeOption(value: Expression): String = {
```

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaTableValueFunctions.scala [hunk 3]
```diff
@@ -110,15 +110,15 @@ trait CDCStatementBase extends DeltaTableValueFunction {
 
   protected def getOptions: CaseInsensitiveStringMap = {
     def toDeltaOption(keyPrefix: String, value: Expression): (String, String) = {
+      val evaluated = DeltaTableValueFunctionsShims.evaluateOptionExpr(value)
```
Comment (cloud-fan) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaTableValueFunctions.scala:113]:
```suggestion
      val evaluated = DeltaTableValueFunctionsShims.evaluateTimeOption(value)
```

**Comment by scottsand-db** [comment]
## Fails Spark Master Tests
https://github.com/delta-io/delta/actions/runs/8983897145/job/24674590454?pr=2999
```
[info] - changes - start and end are timestamps *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "string(current_date())"
[info]   at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
```

```
[info] - changes - start and end are timestamps *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "string(current_date())"
[info]   at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
```

```
[info] - changes - start and end are timestamps *** FAILED ***
[info]   org.apache.spark.sql.delta.DeltaAnalysisException: [DELTA_TIMESTAMP_GREATER_THAN_COMMIT] The provided timestamp (2024-05-07 00:00:00.0) is after the latest version available to this
[info] table (1969-12-31 16:00:02.0). Please use a timestamp before or at 1969-12-31 16:00:02.
[info]   at org.apache.spark.sql.delta.DeltaErrorsBase.timestampGreaterThanLatestCommit(DeltaErrors.scala:1367)
```

## Fails Spark 3.5 Tests
https://github.com/delta-io/delta/actions/runs/8983897146/job/24674591077?pr=2999
```
[info] - changes - start and end are timestamps *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "string(current_date())"
[info]   at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
```

```
[info] - changes - start and end are timestamps *** FAILED ***
[info]   java.lang.NumberFormatException: For input string: "string(current_date())"
[info]   at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
```

```
[info] - changes - start and end are timestamps *** FAILED ***
[info]   org.apache.spark.sql.delta.DeltaAnalysisException: [DELTA_TIMESTAMP_GREATER_THAN_COMMIT] The provided timestamp (2024-05-07 00:00:00.0) is after the latest version available to this
[info] table (1969-12-31 16:00:02.0). Please use a timestamp before or at 1969-12-31 16:00:02.
```

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala [hunk 4]
```diff
@@ -62,6 +65,9 @@ class DeltaCDCSQLSuite extends DeltaCDCSuiteBase with DeltaColumnMappingTestUtil
       case endingVersion: EndingVersion =>
         s"""${endingVersion.value}"""
 
+      case endingTimestampExpr: EndingTimestampExpr =>
+        s"""'${endingTimestampExpr.value}'"""
```
Comment (cloud-fan) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala:69]:
```suggestion
        s"""${endingTimestampExpr.value}"""
```

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala [hunk 5]
```diff
@@ -246,6 +246,15 @@ class DeltaCDCSQLSuite extends DeltaCDCSuiteBase with DeltaColumnMappingTestUtil
           .withColumn("_change_type", lit("insert"))
           .withColumn("_commit_version", (col("id") / 10).cast(LongType))
       )
+      val readDf3 = sql(s"SELECT * FROM table_changes('$tbl', " +
+        "string(now() - interval 1 day), string(now())")
```
Comment (scottsand-db) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala:250]: missing `)` at the end, to close `table_changes`

Comment (cloud-fan) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala:250]:
```suggestion
        "string(now() - interval 1 day), string(now()))")
```

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala [hunk 6]
```diff
@@ -246,6 +246,15 @@ class DeltaCDCSQLSuite extends DeltaCDCSuiteBase with DeltaColumnMappingTestUtil
           .withColumn("_change_type", lit("insert"))
           .withColumn("_commit_version", (col("id") / 10).cast(LongType))
       )
+      val readDf3 = sql(s"SELECT * FROM table_changes('$tbl', " +
+        "string(date_sub(current_date(), 1)), string(now()))")
```
Comment (cloud-fan) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala:250]:
```suggestion
        "string(current_date()), string(now()))")
```

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala [hunk 7]
```diff
@@ -246,6 +246,15 @@ class DeltaCDCSQLSuite extends DeltaCDCSuiteBase with DeltaColumnMappingTestUtil
           .withColumn("_change_type", lit("insert"))
           .withColumn("_commit_version", (col("id") / 10).cast(LongType))
       )
+      val readDf3 = sql(s"SELECT * FROM table_changes('$tbl', " +
+        "string(current_date()), string(now()))")
```
Comment (cloud-fan) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSQLSuite.scala:250]:
```suggestion
      val readDf3 = sql("SELECT * FROM table_changes" +
        s"('$tbl', string(date_sub(current_date(), 1)), string(now()))")
```

**Pull request merged**